File: frontend/src/types/trace.ts
export type TraceType =
  | 'intent_detection'
  | 'plan'
  | 'tool_call'
  | 'tool_result'
  | 'delegation'
  | 'response'

export interface TraceEvent {
  id: string
  session_id: string
  agent_id: string | null
  trace_type: TraceType
  payload: Record<string, unknown>
  tool_name: string | null
  duration_ms: number | null
  status: 'running' | 'success' | 'error'
  created_at: string
}

export interface TraceStreamMessage {
  event: 'trace'
  data: TraceEvent
}

File: backend/schemas/trace.py
"""Pydantic schemas for execution traces."""
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel


class TraceEvent(BaseModel):
    id: UUID
    session_id: UUID
    agent_id: UUID | None
    trace_type: str
    payload: dict
    tool_name: str | None
    duration_ms: int | None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class TraceStreamMessage(BaseModel):
    event: str = "trace"
    data: dict
